keps/sig-node/1592-pod-level-resource-limits/README.md

```diff
@@ -0,0 +1,439 @@
+
+This proposal suggests a middle ground, and suggests a way to make it possible to describe to Kubernetes how to limit the resource consumption of multiple containers in the pod at the Pod level, instead of trying to micro-manage the resource limits on the container level. For workloads where the work performed is burstable, this proposal would make it easier to allow the low-level mechanisms available in the underlying operating-system to manage the resources required for the task.
+
+### Goals
+
+This proposal aims to:
+
+* Allow developers to define resource limits on the pod level in addition to the individual container level.
+* Use as much as possible the Linux kernel's cgroup ability to define limits in hierarchies to enforce the defined limits.
+
+### Non-Goals
+
+*  Providing a general-purpose interface to the full range of possible resource management provided by the Linux cgroup hierarchy.
+
+## Proposal
+
+The Pod QoS enhancement already implemented in Kubernetes manages resources as a hierarchy of cgroups in the following way:
+
+<pre>
+  kubepods
+   |
+   +-- Guaranteed-pod0
+   |   |
+   |   +-- container0 (pause container)
+   |   |
+   |   +-- container1 (first container)
+   |   |
+   |  = ... =
+   |   |
+   |   +-- containerN (N-th container)
+   |
+   +-- QoS CGroup (one of burstable, or besteffort)
+       |
+       \ pod 
+          |
+          +-- container0 (pause container)
+          |
+          +-- container1 (first container)
+          |
+        = ... =
+          |
+          +-- containerN (N-th container)
+</pre>
+
+The proposal in this KEP is to allow users to define resources on the Pod level itself. Both resource requests and limits are allowed. The scheduler
+algorithm will consider the pod-level resources if they exist in preference to the container-level resources. If no pod-level resource section is provided,
+then the scheduler will continue to operate as it does in the current implementation. The QoS definitions will be updated to consider the pod-level resources
+when classifying a pod into one of the QoS levels. If the pod level doesn't include any resource limits, then kubelet will continue to function as it does today.
+
+The current implementation will set the cgroup limits for the memory/CPU resources on the pod level of the hierarchy only in the following cases:
+1. The QoS level is `Guaranteed`. 
+1. The QoS level is `Burstable` and all of the containers specify a limit for the relevant resource. 
+
+In the current implementation pods are considered `Guaranteed` if and only if all of the included containers define a resource request which is equal
+to the resource limit. The current implementation considers a pod to be `Burstable` if at least one container defines resource requests or limits. Any
+pods not matching the `Guaranteed` or `Burstable` QoS definitions are placed in the `BestEffort` QoS level.
+
+This proposal suggests modifying these definitions the following: 
```

Comment:
This section is  awkwardly worded.  I think we can find a better way to express it.  Something like:

"""

For each core resource type (cpu, memory) a pod may have a pod-scoped resource request and/or limit.   Pods which have a defined request or limit for a given resource type are said to "have explicit resources" or "be explicitly resourced".  Pods which DO NOT have a defined request or limit for a given resource type are said to "have implicit resources" or "be implicitly resourced".

A pod is considered `Guaranteed` if, for each core resource type, either of the following is true:
* It is explicitly resourced for that resource type AND the the pod's request is equal to its limit.
* It is implicitly resourced for that resource type AND all containers' requests are equal to their limits.

A pod is considered `Burstable` if,  for each core resource type, any of the following is true:
* It meets the criteria for "Guaranteed" for that resource
* It is explicitly resourced for that resource type AND the pod's request is specified and less than or equal to its limit.
* It is implicitly resourced for that resource type AND all containers' requests are specified and less than or equal to their limits.

A pod is considered `BestEffort` if, for each core resource type:
* It is implicitly resourced for that resource type AND any container's request is not specified

Reply:
We could maybe simplify more.  Could we say that:

"""
For each core resource type, a Pod's resources are either (evaluated in order):
* the values specified in the pod's resources stanza
* the sum of the values specified in all of the pod's containers' resources stanzas, where a single unspecified value means the result is unspecified

A pod is considered `guaranteed` if, for each core resource type, the pod's request equals the limit.
"""